core/src/main/java/jenkins/model/RunIdMigrator.java

```diff
@@ -123,8 +123,10 @@ private void save(File dir) {
         File f = new File(dir, MAP_FILE);
         try (AtomicFileWriter w = new AtomicFileWriter(f)) {
             try {
-                for (Map.Entry<String,Integer> entry : idToNumber.entrySet()) {
-                    w.write(entry.getKey() + ' ' + entry.getValue() + '\n');
+                synchronized (this) {
+                    for (Map.Entry<String, Integer> entry : idToNumber.entrySet()) {
+                        w.write(entry.getKey() + ' ' + entry.getValue() + '\n');
+                    }
```

Comment:
`idToNumber` is a regular `TreeMap` and is written to by other synchronized methods in this class, so to avoid racing with them we should synchronize on `this` as well.